Modules/_ssl.c

```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
     return NULL;
 }
 
+static unsigned int psk_client_callback(SSL *s,
+                                        const char *hint,
+                                        char *identity,
+                                        unsigned int max_identity_len,
+                                        unsigned char *psk,
+                                        unsigned int max_psk_len)
+{
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    PySSLSocket *ssl = SSL_get_app_data(s);
+    if (ssl == NULL || ssl->ctx == NULL) {
+        goto error;
+    }
+    PyObject *callback = ssl->ctx->psk_client_callback;
+    if (callback == NULL) {
+        goto error;
+    }
+
+    PyObject *hint_str = (hint != NULL) ?
+            PyUnicode_DecodeUTF8(hint, strlen(hint), "strict") :
```

Comment:
how safe is it to call `strlen(hint)`? this is raw data being supplied by OpenSSL, does it guarantee a reasonable length and null termination for us? (no hint length is passed, so i'm assuming so)

should the hint be treated as UTF-8?  Does TLS or OpenSSL require that?  It might make more sense to have code use PyBytes instead.  A failed UTF-8 decode is something the other end of the network connection could trigger.  same for the other callback regarding identity.

Reply:
> does it guarantee a reasonable length and null termination for us?

Yes:

> The callback function is given the connection in parameter ssl, a NULL-terminated PSK identity hint sent by the server in parameter hint

The OpenSSL implementation limits `identity` and `identity_hint` string length to `#define PSK_MAX_IDENTITY_LEN 256`, and uses `strndup` to guarantee null termination.
(Oddly this doesn't seem to meet identity length specified in the RFC `opaque psk_identity<0..2^16-1>;`)

> should the hint be treated as UTF-8? Does TLS or OpenSSL require that?

The RFC requires `identity` and `identity_hint` to be encoded as UTF-8:

> 5.1. PSK Identity Encoding
The PSK identity MUST be first converted to a character string, and then encoded to octets using UTF-8.

Although, the OpenSSL implementation doesn't do any UTF-8 validation, it treats what it is given as null terminated bytes. So it would be possible to cause a python exception on the remote side by sending invalid UTF-8 (I assume that is what you would like to avoid).

I think a good approach would be for the C callbacks to just ignore/drop requests with invalid UTF-8. Although this would require a UTF-8 validation function in C that does not raise a python exception (not sure if that exists).

For now I will follow your advice and change these to PyBytes

Reply:
> For now I will follow your advice and change these to PyBytes

@gpshead I've changed my mind on this.
I don't think it makes sense to use PyBytes to handle a UTF-8 string.
This could stay as UTF-8 and call `PyErr_Clear();` if the UTF-8 decode fails and reject the connection.

Could you please elaborate on your concerns if you think this should still be PyBytes?

Reply:
your reasoning makes sense and I like your PyErr_Clear approach when the Decode fails.